### ISSUE TITLE
Add support for newer JDKs

### ISSUE THREAD

**Issue by vlaaad** [cljdoc/cljdoc#275]
Currently cljdoc uses JDK 8 to build library documentation. If library depends on newer JDK, build fails with UnsupportedClassVersionError exception:
 `java.lang.UnsupportedClassVersionError: javafx/beans/value/ObservableValue has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0, compiling:(cljfx/mutator.clj:1:1)`
Example of such failed build: https://circleci.com/gh/cljdoc/builder/1319

**Comment by martinklepsch** [comment]
Hi @vlaaad & thanks for the issue. Supporting newer JDKs would definitely be nice, here's some rough list of steps of what we'll need to do for this (likely incomplete/wrong in places):

- Figure out a way to determine what JDK version to use for analysis
  - Maybe this is something we can read from the Jar or POM file 
  - The less good solution would be to introduce a configuration option where library authors can manually specify what JDK version to use
- Figure out how this maps to CircleCI 
  - Do we need to use a different container image? 
  - Are different java versions available on the CircleCI images? 
  - Can we parameterise what image is used using CircleCI's `BUILD_PARAMETERS`?
- Think about how it affects local analysis and how errors with that are surfaced to users

**Comment by danielcompton** [comment]
* Maybe this is something we can read from the Jar or POM file

Inside the JAR is a `MANIFEST.MF` file. You can pull it out with 

```console
$ jar --extract --file=ring-1.6.3.jar META-INF/MANIFEST.MF
$ cat META-INF/MANIFEST.MF
Manifest-Version: 1.0
Built-By: jim
Created-By: Leiningen 2.7.1
Build-Jdk: 1.8.0_121
Main-Class: clojure.main
```

However, this only tells you which JDK was used to build it, not which Java version it was compiled for (if any compilation took place). It also tells you nothing about the build or compile versions of any dependencies. I _think_ this should mostly be a safe upper bound to pick though, i.e. if something was compiled with JDK 11 but targeted JDK 8, then you should still be able to analyze it with JDK 11. The only wrinkle I can see there is if someone happened to use JDK 11 on a project that wasn't compatible with JDK11. I think, in that case, it would be possible to build the JAR as long as no code was evaluated/compiled as part of building the JAR. That seems like a fairly remote concern though.

To determine the absolute minimum possible JDK version you could use, I think you'd need to [read](https://stackoverflow.com/a/1096159/826486) all of the class files of all of the dependencies and find the minimum class file version used. That would probably be a bit unnecessary though.

* The less good solution would be to introduce a configuration option where library authors can manually specify what JDK version to use

Yeah, I think you should probably be able to work out _a_ valid JDK to use to analyse the JAR by matching whatever was originally used to build the JAR.

> * Do we need to use a different container image?

Probably easiest, although there is [bach](https://github.com/sormuras/bach#install-jdksh) which has a script to install different JDKs. There is also [sdkman](https://sdkman.io/sdks#java) which can install Java.

> * Are different java versions available on the CircleCI images?

Yep, there are lots of variants available. https://github.com/CircleCI-Public/circleci-dockerfiles/tree/master/clojure/images seems to be the best overview I can find.

* Can we parameterise what image is used using CircleCI's `BUILD_PARAMETERS`?

Not sure, I think you probably could but couldn't find it immediately.

* Think about how it affects local analysis and how errors with that are surfaced to users

Not really sure on the fine details, but you could check which version of `java` is available and switch/error out based on that. If you wanted to be super fancy, you could use `java_home` to try and find a compatible JDK:

```console
$ /usr/libexec/java_home -v 11
/Library/Java/JavaVirtualMachines/adoptopenjdk-11.0.1.jdk/Contents/Home
```

**Comment by martinklepsch** [comment]
Thanks for your research on this @danielcompton. 

The manifest file approach sounds promising but [cljfx's jars](http://repo.clojars.org/cljfx/cljfx/1.2.3/) don't seem to include that file. Could be a packaging issue, not sure. cc @vlaaad 

If someone figures out a way to determine the Java version needed for a given jar I think I have a rough idea on how the rest could be done, thanks to the tools Daniel suggested.

**Comment by vlaaad** [comment]
Looks like [depstar](https://github.com/seancorfield/depstar) (which I use for cljfx) [does not generate manifest file](https://github.com/seancorfield/depstar/issues/1). Not a problem though: if cljdoc will look for `MANIFEST.MF` file in the jar, I'll just add it manually as a "resource file", since it's just a file on a known path :+1:

**Comment by seancorfield** [comment]
(just FYI, since `depstar` was mentioned: I have no plans to tackle AOT or manifests in `depstar` and just closed that issue out as "won't do")

**Comment by bbatsov** [comment]
I'm posting a link here to https://circleci.com/gh/cljdoc/builder/7400, as orchard is affected by this as well. 

Orchard has different Java parsers for JDK 8 and for JDK 9+ and selects the right one dynamically, but that the newer one breaks cljdoc's analyzer. I guess it'd be nice to either be able to specify the JDK you want to use or two be able to exclude some namespaces from the analysis.

**Comment by schmee** [comment]
Just chiming in with another repo: https://schmee.github.io/java-http-clj/

Failing build: https://circleci.com/gh/martinklepsch/cljdoc-builder/7411#tests/containers/0

```
Caused by: java.lang.ClassNotFoundException: java.net.http.HttpClient
```

The repo is a wrapper around the HTTP client introduced in Java 11, so it will most definitely not work with a version lower than that 😄 I'll stick to Codox for this particular repo in the meanwhile (very happy with Cljdoc for [Daguerreo](https://cljdoc.org/d/schmee/daguerreo/0.1.0/doc/readme) though!) 👍

**Comment by martinklepsch** [comment]
I recently learned that pom.xml files have a generic properties field that can be used to specify arbitrary things. 

https://maven.apache.org/ref/3.6.1/maven-model/maven.html

```xml 
<properties>
  <cljdocJDKVersion>11</cljdocJDKVersion>
</properties>
```

This is less user friendly than the automatically inferred way of determining JDK versions but its a reasonable option if we wanted to allow users to hardcode a version inside their artifact. 

An automatically inferred way could always be added later on as well.

I guess the biggest outstanding question is how can we parameterize the CircleCI Job to have the correct version of the JDK installed. Probably installing it as part of the build using something like sdkman mentioned above is the best contender right now.

**Comment by rschmukler** [comment]
Just found this issue when trying to build documentation for [teknql/tapestry](https://github.com/teknql/tapestry) which requires a preview build of the loom enabled JVM.

It'd be nice if the solution could allow for any JVM version. What about allowing specifying a custom docker image used to build the documentation in CircleCI inside a `cljdoc.edn`. That'd allow for known (eg. JVM 8, 9, 11, 13) JDKs and, for library authors willing to build their own Docker images, fringe JVM builds.

**Comment by martinklepsch** [comment]
@rschmukler that might be a good approach! Do you want to explore this a bit more / figure out what we'll need to change? Can the docker image be parameterized using CircleCI's build params? 

For reference, here is 

- the current CircleCI config: https://github.com/cljdoc/builder/blob/master/.circleci/config.yml
- and here is how we queue builds that use this config: https://github.com/cljdoc/cljdoc/blob/d6c6fab0ef34dbb2bf3734ba0614466a401c20bb/src/cljdoc/analysis/service.clj#L49-L83

**Comment by rschmukler** [comment]
@martinklepsch awesome, thank you for pointing me in the right direction! I'll try and find some time to take a look ASAP.

**Comment by helins** [comment]
Targeting a specific version is already a good step forwards but I believe we should be able to target a JVM family as well. A practical example is GraalVM, becoming important in the Clojure ecosystem, which has specific APIs not found elsewhere such as the Polyglot API.

Parametrizing the docker image is the most versatile approach. I like sdkman as a developper tool but how would it behave with CircleCI caching? Wouldn't the docker solution be more efficient as well?

**Comment by lread** [comment]
Interesting point about vendors potentially providing different APIs @helins! 
I am guessing these Polyglot APIs can change which specific GraalVM versions?  

Have we thought about maintenance, support and security issues around letting folks specify any docker image?
Any concerns?  

I can work on this issue if we can figure out what we want to do.

I think the latest idea was specifying a JDK, or alternatively, a specific docker image in `cljdoc.edn`?
I personally think `cljdoc.edn` is a more obvious (eg. easy for doc authors) choice than a manifest file or pom file.

**Comment by lread** [comment]
Thanks for continuing the conversation @helins! 

Yeah, I am not sure what security risks would be introduced by allowing cljdoc authors to choose any available docker image, but mainly it would add an aspect the cljdoc dev team did not previously have to think of.

What do folks think about this?:
1. For repeatable builds we probably (?) would want to stick with specific JDK versions.  
For example, a cljdoc user might be able to choose AdoptOpenJDK v11, but cljdoc would use a specific backend configured version. The cljdoc team would manually update this configuration periodically.

2. I see our choice of JDK installation tool as important, but also as an implementation detail.
    If we choose to go with SDKMAN!, [I think it recently added support for installing older versions that it does not list anymore](https://github.com/sdkman/sdkman-db-migrations/issues/170), so that might help us.  

3. I also think we could start with a single vendor only (maybe AdoptOpenJDK) and add support for other vendors as requested by the Clojure community. Less knobs to twirl means less flexibility but also less support work for the cljdoc team. The idea is that we would add flexibility as it is actually needed by the Clojure community.

   If this makes sense to people, we would likely only allow something like `jdk-version` to be specified by cljdoc authors, and initially, it would only specify the major version.

4. If we go with specifying JDK selection in `cljdoc.edn`, [we should update our `verify-cljdoc-edn` script to also validate our new config options](https://github.com/cljdoc/cljdoc/blob/master/doc/userguide/for-library-authors.adoc#configuring-articles).  

5. Something else to think about is that JDK versions eventually go out of support.
It might be that in the future cljdoc might reject a JDK version that it initially accepted, at least for new doc builds?
I dunno.

Continued discussion/ideas most appreciated and welcome!

**Comment by helins** [comment]
I don't mean to be too insistent on SDKMAN, there are probably things I haven't considered. It's just that it looks like it requires "only" some initial work with very little maintaining afterwards. Maintaining a single docker image with SDKMAN as opposed to adding burden on maintainers having to constantly maintain a set of backends. I assume no one wants extra work unless strictly necessary. So, commenting on your points:

1. Check, choosing a JVM means specifying a complete identifier such as "20.2.0.r11-grl" from `sdk list java`.

2. I believe it is quite rare tor JVM features to be totally removed. From what I've seen, things usually stay deprecated for years and might no even get fully removed. Hence I believe it is more important being able to analyze against newer version than older ones, since we might need to access new features. So I feel less concerned about SDKMAN removing older versions, however I don't know how volatile it is. Plus, it is fair to assume that analysis performed at some moment is meant for a project requiring common JVMs at that moment. Not older versions that wouldn't even be listed in SDKMAN (yet again, don't know how volatile it is). I wouldn't release a Clojure library today that exclusively needs Java 6.

3. Wouldn't be a problem, as per point 1. In reality, the main alternative vendor I envision is GraalVM since it is quite specific and also becoming important for Clojure. I can't think of a vendor not available on SDKMAN that someone would want.

4. /

5. As per point 2, that might happen. However, also as per point 2, analysis will be performed at a moment where the chosen JVM is still available. So, what are the chances of having to re-run analysis so far in the future that it requires an unavailable JVM, without modyfing the project?

**Comment by rschmukler** [comment]
@helins it doesn't look like sdkman would be able to handle my use case of being able to use fringe JDKs (such as the experimental JDK that's built and hosted on the [loom project](http://jdk.java.net/loom/).

Another example I can think of of people wanting to be able to use a custom JDK beyond what Sdkman provides is JDKs targetting embedded devices (such as Oracle Java Embedded, or Azul Zulu Embedded) or commercially enhanced JDKs (eg. Azul Zing).

Not saying that these use-cases are common enough to warrant supporting, but it is worth considering since if we're solving it, it would be nice if our "escape hatch" was flexible enough to handle all use cases.

**Comment by lread** [comment]
@helins I'm suggesting that we abstract away from any particular implementation choice and consider adding JDK distributions on an as requested basis.  

@rschmukler Sticking with the mainstream open source friendly JDKs probably makes sense for cljdoc.  
We probably don't want to have to figure out, for example, if a fringe JDK is legal for cljdoc to use.
Or is that really a valid concern?
I figure people who want to do their own very very special thing can always use Codox and host their own docs.

@martinklepsch When you have some time and interest, would be great to hear your thoughts.
I am certainly swaying toward a conservative approach here, but maybe you have other ideas.

**Comment by lread** [comment]
I wonder if I was just being a bit lazy when I discounted letting cljdoc authors optionally specify a docker image.
Me not wanting to think about it doesn't make it a bad idea, does it? 🙂 

The CircleCI analysis job is, by design, isolated.
All it does it generate an edn file.

Cljdoc then picks up that edn file, validates it with spec, and loads it into its database.

Random thoughts/questions:
1. There is already a barrier to entry to having cljdoc analyze your library; it must be up on clojars (or some other maven repo). The benefit to this barrier is that random bad actors are less likely to experiment with cljdoc.
2. Would/could someone generate with a custom docker image that would harm cljdoc?  Would we need to add any more validation here?
3. CircleCI has been generous in blessing cljdoc's use of its service. Would they have any complaint to us opening it up to use an arbitrary docker image?
4. What would it mean if a docker image tag pointed to a docker image that once worked, but no longer works? I suppose that might mean historical projects with custom docker images might be more difficult to reconstitute.  Why would we want to reconstitute? Maybe cljdoc suffers a catastrophic data loss.  This could simply be a known potential cost of using custom docker images.
5. Is a custom docker image a support nightmare? Probably not. Most folks would use the default image or specify an alternate recommended docker image. I expect folks choosing a docker image outside the recommended set would be very small.
6. Would cljdoc be technically impacted by a custom docker image?
    1. I do know that cljdoc currently times out an analysis job after 10 minutes.  This value might have to be increased if a custom image needs more time to get analysis done.
    2. Could multiple hung/failing analysis jobs negatively impact cljdoc?

**Comment by martinklepsch** [comment]
> Yeah, I am not sure what security risks would be introduced by allowing cljdoc authors to choose any available docker image

I think the risks are somewhat limited in that the CI job doesn't have write access to the cljdoc database. Docker images seem like a good (low) level of abstraction. We could still whitelist what docker images are allowed and which aren't. That said I'm not sure how easy using custom docker images with CircleCI would be. Last time I looked it wasn't obvious how I could parameterize a build so that it uses a different docker image. Probably there's a way 😄 

Nextjournal has offered compute resources for cljdoc on a few occasions, maybe that'd be an option if CircleCI doesn't work for this scenario (we could still keep the 99% default config projects on the CircleCI analyzer). 

> Would/could someone generate with a custom docker image that would harm cljdoc? Would we need to add any more validation here?

There's one "attack vector" I can think of where a bad actor could delete the data stored on cljdoc for a single artifact at the time. Overall I'm not too worried about that type of stuff since data can be recovered and incentives for this type of attack are low. 

> Maybe cljdoc suffers a catastrophic data loss. 

A good reminder for me to run `./ops/backup.sh` :joy:

**Comment by lread** [comment]
Thanks for chiming in @martinklepsch, much appreciated.

I shall look into possibilities.  The v2 CircleCI API has some new parameter support, I'll play around and report back.

**Comment by lread** [comment]
Yeah, the CircleCI v2 API along with a v2.1 config look promising if we feel good about going with docker images.

Just tested with a modified forked cljdoc/builder, relevant bits of config.yml:
```yaml
version: 2.1
parameters:
  image-name:
    type: enum
    default: "circleci/clojure:openjdk-11-tools-deps"
    enum: ["circleci/clojure:openjdk-11-tools-deps", "circleci/clojure:openjdk-8-tools-deps"]

jobs:
  build:
    docker:
      - image: << pipeline.parameters.image-name >>
.
.
.
      - run:
          name: Tools versions
          command: |
            clojure -Sdescribe
            java -version
```
(Note I didn't use an `enum` parameter for any great reason, just playing at this point).

Parameters are passed in like so when triggering a pipeline:
```Shell
curl -X POST https://circleci.com/api/v2/project/gh/lread/builder/pipeline \
 --header 'Content-Type: application/json' \
 --header 'Accept: application/json'    \
 --header "Circle-Token: $CIRCLE_TOKEN" \
 -d '{ "branch": "lread-testing123","parameters":{"image-name":"circleci/clojure:openjdk-8-tools-deps" }}'
```
...results in Tools versions step output of:
```
#!/bin/bash -eo pipefail
clojure -Sdescribe
java -version

{:version "1.10.3.822"
 :config-files ["/usr/local/lib/clojure/deps.edn" "/home/circleci/.clojure/deps.edn" ]
 :config-user "/home/circleci/.clojure/deps.edn"
 :config-project "deps.edn"
 :install-dir "/usr/local/lib/clojure"
 :config-dir "/home/circleci/.clojure"
 :cache-dir "/home/circleci/.clojure/.cpcache"
 :force false
 :repro false
 :main-aliases ""
 :repl-aliases ""}
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
```

Triggering again without parameters specified will choose the default parameter:
```Shell
curl -X POST https://circleci.com/api/v2/project/gh/lread/builder/pipeline \
 --header 'Content-Type: application/json' \
 --header 'Accept: application/json'    \
 --header "Circle-Token: $CIRCLE_TOKEN" \
 -d '{ "branch": "lread-testing123"}'
```
...results in Tools Versions step output of:
```
#!/bin/bash -eo pipefail
clojure -Sdescribe
java -version

{:version "1.10.3.822"
 :config-files ["/usr/local/lib/clojure/deps.edn" "/home/circleci/.clojure/deps.edn" ]
 :config-user "/home/circleci/.clojure/deps.edn"
 :config-project "deps.edn"
 :install-dir "/usr/local/lib/clojure"
 :config-dir "/home/circleci/.clojure"
 :cache-dir "/home/circleci/.clojure/.cpcache"
 :force false
 :repro false
 :main-aliases ""
 :repl-aliases ""}
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment 18.9 (build 11.0.10+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.10+9, mixed mode, sharing)
```

We should move to the CircleCI v2 API if we can anyway, as the v1.1 API is now deprecated I think.

I'll continue to experiment, at the very least it will give us an idea of what doors are open to us.

**Comment by lread** [comment]
An interesting aside: Things have changed since this issue was first raised in February of 2019.

I just noticed that our [CircleCI analysis job is using docker image `circleci/clojure:tools-deps`](https://github.com/cljdoc/builder/blob/56a713989e29398e8a36a02c3d405299dd8d9cc2/.circleci/config.yml#L12).
This once brought in JDK8 but now brings in JDK11.
[Seems the default JDK change was made in September of 2019](https://github.com/Quantisan/docker-clojure/pull/70).

At the time of this writing, it seems cljdoc has been using JDK11 for analysis jobs for about a year and a half.  
I don't know that anybody noticed this. 
This might mean that JDK11 is not problematic for cljdoc analysis jobs - unless people have been suffering in silence?

This doesn't necessarily mean adding support for alternate JDKs has no value.
@rschmukler, you seem to have a project that fails analysis on cljdoc, but would pass if you could choose a different JDK.
Are there other real-world non-hypothetical projects that we know of that are failing to analyze on cljdoc due to an inability to select the JDK?

**Comment by lread** [comment]
Actually @vlaaad, I see that https://cljdoc.org/d/cljfx/cljfx/1.7.13 is still failing to api analysis today.
Would an alternate JDK solve the problem for you?

**Comment by helins** [comment]
@lread  Regarding uses cases.

Given that there is quite some work being currently done in data science, number crunching, and interoperability with other languages, I mainly see 2 problems that will hinder those efforts:

1) Not being able to analyze libraries leveraging Graal's polyglot API (embedding Python, WebAssembly, ...) (important)
2) Not being able to leverage JDK 16 features for calling native code (less important, still experimental)

Of course, that's about an obscenely small percent of Clojure libraries in comparison to what Cljdoc holds. But I would love to see Clojure becoming stronger in those areas so that it can truly become "the language of languages". That is, if I am not dreaming too big ;)

**Comment by lread** [comment]
Thanks @helins, much appreciated.  Sounds like you are pretty certain you'd make use of an alternate JDK with cljdoc.

**Comment by KingMob** [comment]
Just ran into this issue today. I upgraded [True Grit](https://github.com/KingMob/TrueGrit), my Resilience4j wrapper lib, to use the new 2.0.0 version of R4j, and cljdoc broke, because 2.0.0 now requires Java 17.

The last comment was over a year ago...any news on this issue?

**Comment by KingMob** [comment]
Hmm, well, this is a problem that will only get worse over time, not better.

It seems there was a lot of discussion centered around supporting esoteric JDKs, dealing with Circle Docker images, etc. 

Instead of letting the perfect be the enemy of the good, what about starting with a shortlist of allowable JDKs, like the ones with LTS support, and letting people pick one with a flag in cljdoc.edn? That would cover most use cases, I'd think. 

I don't doubt there are more sophisticated ways to handle this issue, but I wonder if it's gotten bogged down in analysis paralysis.

**Comment by lread** [comment]
The concern is that we'll have to support this feature forever. 
And we want to make sure what we implement is future-proof.

But yeah, starting small and vague is probably safe.

Perhaps we could start with `:cljdoc/jdk-version` (in `doc/cljdoc.edn`) and only support only values `8` and `17`.
Cljdoc analysis would choose the specific version (likely the latest release) and jdk vendor (probably temurin, we'll see).

And then we can react to what folks request down the road.

Another thing that will help is supporting static analysis, which I took some notes on a while ago in #543.

**Comment by esuomi** [comment]
An alternative to consider: Tools such as [jenv](http://jenv.be/) use a `.java-version` file at the root of the project to describe the version of JDK the project relies on. Since this allows for ecosystem friendly, build tool independent way of expressing the desired version(*) I'd rather see that being used instead of putting it into specific tool configuration.

*) There's the obvious thing that for some the exact distribution of JDK (eg. Azul, Corretto, Graal etc.) might be important, but maybe that should be the second version of this support, start with the easier thing first... :)

**Comment by lread** [comment]
It is so nice that this issue is getting some interest from the community of late!
Thanks folks, the issue that squeaks is typically the one that gets attention.

@esuomi I think your idea is interesting. But... different tools use different conventions.  For example sdkman uses `.sdkmanrc`. And... tools and conventions can change.

So, I'm thinking that depending on our own config might not be as convenient, but it keeps things clear and disentangled. And is probably a reasonable (and simple) starting point.

**Comment by martinklepsch** [comment]
We could also look at different files with a clear order of precedence. 

But I think the challenge still is less the question of how to specify the version and rather about "how do we run build jobs with specific JDK versions?"

I think if we had a prototype of our CircleCI `AnalysisService` that allows running cljdoc analysis under a specific JDK then the question of how to specify this version would fall into place :)

**Comment by lread** [comment]
Great to see you have some time for cljdoc today @martinklepsch! ❤️ 

Yeah true. But the interface (config) is still very important. If we decouple it from the implementation this frees us to change the implementation when we want/need to.

For the implementation, I did experiment a while ago with telling CircleCI to use different docker images for different SDKs. If I remember correctly, I switched to the v2 CircleCI API and used parameter passing.

But... This strategy does not help when running locally.
We can run locally from source or docker image.
When running from source we could argue that the dev could just manually setup the appropriate JDK. 
When running locally from docker image, the analysis is run within the docker image. We probably want to somehow select the configured JDK for the analysis run. 
Not that our cljdoc docker image is currently using JDK 17.

So maybe a tool like sdkman (or similar) would be a better choice.
Worth an experiment/investigation anyway.

I can play around with this in a bit, unless somebody else is excited to give it a try.

**Comment by martinklepsch** [comment]
I'd be fine with restricting this during local development and throwing an exception when trying to analyze a project that requires a different JDK version than the one cljdoc is running in? 

Maybe making that decision would allow for progress on this in a way where we unblock 99% of users (i.e. those that don't run cljdoc locally).

> Great to see you have some time for cljdoc today @martinklepsch! ❤️

Thanks, great to jump in here and there & thanks for your pushing on so many of these issues ❤️

**Comment by lread** [comment]
> I'd be fine with restricting this during local development and throwing an exception when trying to analyze a project that requires a different JDK version than the one cljdoc is running in?
> 
> Maybe making that decision would allow for progress on this in a way where we unblock 99% of users (i.e. those that don't run cljdoc locally).

That sounds reasonable to me. 
Dunno, maybe log a warning instead of throw an error?

I'll likely give sdkman a quick whirl anyway just to understand more how/where it would fit in for local and ci use cases. 

But if I get bogged on that, I can fall back to CI solution I experimented with a while ago.

**Comment by martinklepsch** [comment]
> But if I get bogged on that, I can fall back to CI solution I experimented with a while ago.

If you want we could pair on it, could be fun 🙂 Not that you'd need me, more as a random opportunity to hang out 😄

**Comment by KingMob** [comment]
Weirdly, I ran cljdoc Docker locally on True Grit and had no problems whatsoever. It was only when I deployed, that the _production_ cljdoc complained. 

I just spun the container up again to check, `docker exec`ed a shell, ran `java -version`, and got my local Java version, Temurin-17.0.4.1+1.

I'm not explicitly doing anything with java, just `docker run --rm --publish 8000:8000 --volume /Users/matthew/.m2:/root/.m2 --volume /tmp/cljdoc:/app/data cljdoc/cljdoc`

Not sure what's happening, but maybe local Docker has already been running later/local java versions, and is thus a non-issue? 🤞🏻

**Comment by lread** [comment]
Thanks @KingMob, for your persistence here.
Yeah you are right, there is an inconsistency in JDK versions when running locally vs prod.

In production we do not run API analysis on the same machine as the cljdoc web server.
It is farmed out to CircleCI.
We do this this because API analysis runs unknown code and is technically unsafe.

When running locally this isolation does not happen.
Analysis is spawned on the cljdoc web server machine.

When running locally from source, this will be your dev machine.
When running locally on docker, this will be the docker server container, which as you've learned, is running JDK17.

The CircleCI job is currently running JDK11.

A simple short-term thing I can do is bump the CircleCI job to use JDK17.
I'm not 100% sure but I think that a move from JDK11 to JDK17 won't negatively impact folks. I think it will help more than it hurts.

That probably means we won't bother with JDK selection through config at this time.
We'll wait for more folks to squeak.
I'm thinking the static analysis feature might be a better investment here. 
Static analysis is JDK version agnostic.

**Comment by lread** [comment]
> > But if I get bogged on that, I can fall back to CI solution I experimented with a while ago.
> 
> If you want we could pair on it, could be fun 🙂 Not that you'd need me, more as a random opportunity to hang out 😄

Sounds like fun, but this old boy has never e-paired before. So setup might be a challenge?
Something technically less challenging, but fruitful might be to review current issues together. Maybe we can close out some old stale ones and rediscover interesting ones.

**Comment by martinklepsch** [comment]
I think just bumping to JDK 17 is an interesting idea we could try.

**Comment by lread** [comment]
I can do this sometime today.  I will also get us off that CircleCI legacy docker image and get rid of this warning:
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/967328/203806419-154993d7-1a28-4223-a89b-ec328f10785f.png">

**Comment by lread** [comment]
Note for the future: an example breakage from upgrading a JDK: moving to JDK19 introduces a Clojure interop issue for Thread/sleep. Example from etaoin: https://github.com/clj-commons/etaoin/issues/503

**Comment by lread** [comment]
Feedback from Slack: for jdk19 it will probably be a good idea to automatically include `--enable-preview` java opt. Folks will likely be using preview features in jdk19.
